### PR TITLE
calculations client side

### LINKED ISSUE #2876: High priority - Delay of the validation and graph 1a/1b
@minotogna 

table 1a: it take more than 10 mins to show the errors

### LINKED ISSUE #2881: Table 2a : unable to delete values
@minotogna 

I try to remove the values, but cannot do it at least immediately, and sometimes it's difficult to save the data. PS: Replace a value is working.
[19acfa28-749b-4c79-b55c-8646df592729.webm](https://github.com/openforis/fra-platform/assets/99216009/5f31fb33-d2d5-475d-8182-e6c50810092f)

### LINKED ISSUE #2889: Error message
@minotogna 

Not sure why there are these error messages(**in review**): I use China as a trial and login as NC all the time, plz double-check when possible

### LINKED ISSUE #2891: High priority tab 2b
@minotogna 

No data can be entered for the column **Common name** and data cannot be saved for **Million m³**

[e0229b85-e804-4652-a2ff-bf4dd7ff308f.webm](https://github.com/openforis/fra-platform/assets/99216009/31796562-af65-4627-944b-2ac9f2451f88)

### PR DESCRIPTION
**Pull request by minotogna**
Resolves #2881 
Resolves #2876 
Resolves #2891 
Resolves #2889

### CODE REVIEW

File: src/client/pages/AssessmentSection/DataTable/hooks/useData/useData.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        const formula = Cols.getCalculateFn({ cycle, row, col })
+
+        if (colName && formula) {
+          const paramsValue = { assessmentName, cycleName, countryIso, tableName, variableName, colName, data }
+          const value = RecordAssessmentDatas.getNodeValue(paramsValue)
+
+          if (Objects.isEmpty(value) || value.calculated) {
+            const paramsCalculate = { assessment, countryIso, cycle, data, colName, row, formula }
+            const raw = ExpressionEvaluator.evalFormula<string | unknown>(paramsCalculate)
+            // console.log('2 ---- calculating using data ', data)
```
Comment (sorja) [inline src/client/pages/AssessmentSection/DataTable/hooks/useData/useData.ts]: some redundant comments here

File: src/client/pages/AssessmentSection/DataTable/hooks/useGetData/tableDependencies.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { Assessment, AssessmentName, Cycle, CycleName, TableName } from 'meta/assessment'
+
+export type TableDependency = {
+  assessmentName?: AssessmentName
+  cycleName?: CycleName
+  tableName: TableName
+}
+
+export class TableDependencies {
```
Comment (sorja) [inline src/client/pages/AssessmentSection/DataTable/hooks/useGetData/tableDependencies.ts:9]: is class preferred here?

Comment (minotogna) [reply src/client/pages/AssessmentSection/DataTable/hooks/useGetData/tableDependencies.ts:9]: what do you mean, sorry ?

File: src/client/pages/AssessmentSection/DataTable/hooks/useData/useData.ts (hunk 1 above)
Comment (minotogna) [reply src/client/pages/AssessmentSection/DataTable/hooks/useData/useData.ts]: yo, pr is still draft

File: src/server/controller/cycleData/updateDependencies/updateNodeDependencies.ts [hunk 3]
```diff
@@ -5,16 +5,19 @@ import { BaseProtocol } from 'server/db'
...
+ * @deprecated
+ */
 export const updateNodeDependencies = async (
   props: Omit<PersistNodeValueProps, 'value'> & { sourceNode?: NodeUpdate; isODP?: boolean },
   client: BaseProtocol
   // ): Promise<{ nodeUpdates: NodeUpdates; validations: NodeUpdates }> => {
 ): Promise<{ nodeUpdates: NodeUpdates }> => {
   // const { isODP, sourceNode } = props
-  const { sourceNode } = props
+  // const { sourceNode } = props
```
Comment (sorja) [inline src/server/controller/cycleData/updateDependencies/updateNodeDependencies.ts:17]: can this be removed?

**Comment by sorja** [comment]
Looks good, good job and the performance is much better!

**Comment by minotogna** [comment]
> Looks good, good job and the performance is much better!

Thank you 😊 . the best improvement is the performance , I agree

**Pull request merged**